Platform attributes: optimize wording and format

ThomasVitale opened the pull request:
- Format the attributes to combine headline and explanation for better readability and clearer understanding.
- Combine self-service and automation since they share the explanation and foundation.
- Add "secure by default" which was discussed in the gdoc but not ported over to GitHub

ThomasVitale commented:
@joshgav I reorganised the content of the section to combine headline and explanation for a better flow and clarify. I have also added the "secure by default" point we discussed a while ago on GDocs. What do you think?

abangser commented on platforms-whitepaper/v1alpha1/paper.md:
```diff
@@ -154,62 +154,55 @@ of capabilities, platforms make it easy for their users to deliver valuable prod
+   and it should be designed and evolved based on those requirements, similar to any
+   other software products.
+1. **Common requirements**. A platform should provide the necessary capabilities to
+   support the most common use cases across stream-aligned teams, and prioritize those
```
> stream-aligned teams

Do we define this anywhere? I love Team Topologies, but this is not a globally known concept (yet!).

*Note: Of course this is a question for the paper and not this specific PR, just want to call attention as I spotted it here!

roberthstrand replied:
Do we feel that the wording is so self-explanatory that we could use it even if the reader doesn't know Team Topologies?

ThomasVitale replied:
Good point! I wasn't sure about this, and then I saw we were adding the definition in https://github.com/cncf/tag-app-delivery/pull/298/files and went with "stream-aligned teams". But I agree that "product teams" would be better in this context since it's more self-explanatory. I'll change that.

ThomasVitale replied:
I have replaced "stream-aligned teams" with "product teams".

abangser commented on platforms-whitepaper/v1alpha1/paper.md:
```diff
@@ -154,62 +154,55 @@ of capabilities, platforms make it easy for their users to deliver valuable prod
+   support the most common use cases across stream-aligned teams, and prioritize those
+   over more specific capabilities that are only used by a single team to maximize
+   the value delivered.
+1. **User experience**. A platform should offer its capabilities through consistent
+   interfaces and focus on the user experience. Users should be able to request, manage,
+   and observe such capabilities using a combination of GUIs, APIs, command-line tools,
+   IDEs integrations, and portals, depending on their needs. For example, a platform
```
> Users should be able to request, manage,
> and observe such capabilities using a combination of GUIs, APIs, command-line tools,
> IDEs integrations, and portals, depending on their needs

This seems to challenge our points about thinnest viable platform, though I think the intention is spot on. I wonder we can shorten to just the first sentence? Or if we can some how couch the remaining explanation as a bigger endeavour than some platforms need?

roberthstrand replied:
I agree. I think the first sentence is enough to get the point across, while listing everything one could build for a better DX can quickly become bloated.

ThomasVitale replied:
Should I keep the first sentence, remove the second one, and then leave the example? Or would you prefer leaving the example off as well?

abangser replied:
I think the last sentence is great where it brings to life the concept of meeting platform users where they are. Maybe a trimmed down version would be:

> 1. **User experience**. A platform should offer its capabilities through consistent
   interfaces and focus on the user experience. Platforms should endeavor to meet their users where they are which may mean different interfaces such as IDEs; command-line tools, chat bots, or web portals.

joshgav replied:
+1 to the change and something like Abby's refactor, but Abby I'm not sure if you're suggesting removing the following sentence with an example or not? I'm okay with keeping it, it gives a bit of color; but okay with removing it too if we feel it's repetitive.

ThomasVitale replied:
I have rephrased the paragraph based on your suggestions. What do you think?

abangser replied:
love it Thomas!

abangser commented on platforms-whitepaper/v1alpha1/paper.md:
```diff
@@ -154,62 +154,55 @@ of capabilities, platforms make it easy for their users to deliver valuable prod
+   as a golden path.
+1. **Self-service**. A platform should be self-serviceable. Users must be able to request and
+   receive capabilities autonomously and automatically. This property is key to allowing a platform
+   team to enable multiple stream-aligned teams at scale. The platform capabilities should be
```
This sentence sounds a bit repetitive as is. With this edit I am trying to highlight the difference of handling current needs and scaling in the future.

```suggestion
   team to enable multiple stream-aligned teams and scale as needed. The platform capabilities should be
```

abangser reviewed (commented):
I think this is a really nice change.

One byproduct of this formatting is that it comes across as more a definitive list with specific, definable, labels. The clarity and succinctness that the bolded list headers brings, also breeds an expectation of accuracy and completeness.

I will think on this a bit myself as I am not sure which intent I would have expected or preferred!

roberthstrand commented:
@ThomasVitale, could you run through your commit and sign-off, so that the DCO check works out? Details: https://github.com/cncf/tag-app-delivery/pull/305/checks?check_run_id=11384801820

joshgav reviewed (changes requested):
This is a great update, thank you @ThomasVitale. I like how you merged the contents as headings, it feels like the reader will follow this better.

I lean to agree with @abangser that we shouldn't introduce the term "[value] stream" at this point; let's see if we can use "product" or "application" instead, to reduce cognitive load on our readers! :laughing:

I'm also going to think about the phrases in the headings like "Cognitive load" and "Common requirements", if there might be a way to make them more expressive. :thinking: 

Thanks!

ThomasVitale commented:
@roberthstrand I should have fixed the signoff now

ThomasVitale commented on platforms-whitepaper/v1alpha1/paper.md:
```diff
@@ -154,62 +154,55 @@ of capabilities, platforms make it easy for their users to deliver valuable prod
-database themselves.
+After defining what a platform is and why an organization might want to build one,
+it's relevant to identify what are the key attributes that affect the success
+of a platform.
```
I have updated that paragraph as suggested.

ThomasVitale commented:
@abangser @joshgav I thought a bit about the impression of the bullet list with headings in bold. I don't know how much it helps, but I replaced the phrasing "THE key attributes" with "SOME key attributes". Perhaps we can extend the initial paragraph to make it even clearer that these are just examples of attributes and not a definitive list?

Also, as @joshgav pointed out, the "Common requirements" and "Cognitive load" headings were not very expressive. I thought about it and want to propose removing the first one to make it part of the "Platform as a product" bullet, since that's basically part of it: good product management is also about focusing on the high-value requirements first that will benefit the most users. I also replaced "Cognitive load"  with "Reduced cognitive load for users". What do you think?

abangser commented:
I really like the updates you've made @ThomasVitale 👍

ThomasVitale commented:
Thanks a lot, @abangser!

joshgav commented:
Thanks @ThomasVitale! The adjustment from "the" to "some" makes sense, thanks. I'm a bit less sure about integrating "platform as product" with "focus on common requirements" but am willing to go with it.

Re mentioning other attributes/objectives/values expected from internal platforms, we do want this paper to evolve based on feedback. Maybe a good approach is to couch the whole paper when released in a call to action to give us feedback and help us iterate, let's discuss in our next call.

muemich reviewed (changes requested):
Add "Accelerate onboarding of current developers to new projects and teams by sharing the same platforms between them."
To: https://github.com/cncf/tag-app-delivery/pull/305/files#diff-defb6c1a2cbd28999ffdb84871bd5e6b34547d55f8e3bb88deebe835b829eee2L63

joshgav commented:
Thank you @ThomasVitale! @muemich can you rebase your proposal on this latest change? Thanks!

muemich commented:
Sure, will take a bit as I'm currently on the road.
